tremor-script/src/srs.rs

```diff
@@ -13,13 +13,141 @@
+impl Debug for Deploy {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        self.script.fmt(f)
+    }
+}
+
+/// Captures deployment ready artefacts for one troy unit of deployment
+#[derive(Debug)]
+pub struct UnitOfDeployment<'script> {
+    /// Instances for this deployment unit
+    pub instances: HashMap<String, Box<ast::deploy::CreateStmt<'script>>>,
+}
+
+impl Deploy {
+    /// Provides a Graphviz dot representation of the deployment graph
+    #[must_use]
+    pub fn dot(&self) -> String {
+        self.script.dot()
+    }
+
+    /// borrows the script
+    #[must_use]
+    pub fn suffix(&self) -> &ast::Deploy {
+        &self.script
+    }
+    /// Creates a new Payload with a given byte vector and
+    /// a function to turn it into a value and metadata set.
+    ///
+    /// The return can reference the the data it gets passed
+    /// in the function.
+    ///
+    /// Internally the lifetime will be bound to the raw part
+    /// of the struct.
+    ///
+    /// # Errors
+    /// errors if the conversion function fails
+    pub fn try_new<E, F>(mut raw: String, f: F) -> std::result::Result<Self, E>
+    where
+        F: for<'head> FnOnce(&'head mut String) -> std::result::Result<ast::Deploy<'head>, E>,
+    {
+        use ast::Deploy;
+        let structured = f(&mut raw)?;
+        // This is where the magic happens
+        // ALLOW: this is sound since we implement a self referential struct
+        let structured = unsafe { mem::transmute::<Deploy<'_>, Deploy<'static>>(structured) };
+        // This is possibl as String::into_bytes just returns the `vec` of the string
+        let raw = Pin::new(raw.into_bytes());
+        let raw = vec![Arc::new(raw)];
+        Ok(Self {
+            raw,
+            script: structured,
+        })
+    }
+
+    /// Extracts SRS statements
+    ///
+    /// This clones all statements
+    #[must_use]
+    pub fn extract_stmts(&self) -> Vec<DeployStmt> {
+        // This is valid since we clone `raw` into each
+        // self referential struct, so we keep the data each
+        // SRS points to inside the SRS
+        self.script
+            .stmts
+            .iter()
+            .cloned()
+            .map(|structured| DeployStmt {
+                // THIS IS VERY IMPORTANT (a load bearing clone)
+                raw: self.raw.clone(),
+                structured,
+            })
+            .collect()
+    }
+
+    /// Analyses a deployment file ( troy ) to determine if the
+    /// specification is deployable.
+    ///
+    /// This analysis will check that flow definitions and instances
+    /// are correctly defined based on static compile time checks.
+    ///
+    /// Runtime checks are not performed.
+    ///
+    /// # Errors
+    /// If definitions are incomplete or invalid and instances
+    /// are not deployable based on static analysis
+    ///
+    pub fn as_deployment_unit(&self) -> Result<UnitOfDeployment> {
```

Comment:
I don't think this is sound. since we clone out of `.script` which was transmuted into `'static` the result here will (incorrectly) claim to be `UnitOfDeployment<'static>` leading to the chance of referencing memory that is no longer owned.

To resolve this `UnitOfDeployment` will need to be lifetime less and have a `Vec<Arc<Pin<...>>>` that gets cloned from `self.raw`

Reply:
As above - lets document these carefully!

Reply:
unresolving this since it's not resolved

Reply:
`UnitOfDeployment` is now free of lifetimes so I think this is resolved as a consequence